alternator/streams.cc

```diff
@@ -486,47 +480,130 @@ future<executor::request_return_type> executor::describe_stream(client_state& cl
         auto i = topologies.begin();
         auto e = topologies.end();
         auto prev = e;
-        auto threshold = db_clock::now() - confidence_interval(db);
         auto shards = rjson::empty_array();
 
         std::optional<shard_id> last;
 
+        // filter out cdc generations older than the table or now() - 24h
+        auto low_ts = std::max(as_timepoint(schema->id()), db_clock::now() - 24h);
+        while (i != e) {
+            auto j = i;
+            if (++j != e && j->first < low_ts) {
+                i = j;
+                continue;
+            }
+            // next is younger.
+            break;
+        }
```

Comment:
Why not? Easier to read, easier to change. And yes, this is literally lower_bound, however, since we're using map iterators (forward only), the normal implementation of lower bound (which is using binary search internally) will be much more expensive in most cases. Not that it matters much, but it is still quite unnecessary.

Reply:
`map::lower_bound` takes a key, not an iterator, and is logarithmic in the size of the container.

Reply:
Ooh, newfangled c++14 methods. Forgot those were added.

Reply:
That's why I gave you a link, you could have clicked on it :(

Reply:
Old eyes do not see well. Thought you linked to std::lower_bound, because you said "stl algo", not "std::map member".

Reply:
Unfortunately, neither of the std::algo or member functions works here, because it is neither upper or lower bound as defined, it is the last element _not greater_ (i.e. equal or lower), which either means convoluted reverse iterators + inverted comparator etc, or similar. Imho much less readable than a simple loop.

Reply:
But I guess we can make it with an additional `--`... if it makes you happy.

Reply:
I thought it's the last element which is *lower*, because the code breaks when the next element is `>= low_ts`. You're saying *lower or equal*, in which case it should break when next element is `> low_ts`, i.e. this line:
```
 if (++j != e && j->first < low_ts) {
```
should instead be
```
 if (++j != e && j->first <= low_ts) {
```

In any case:
1. if you meant *lower*, you could do `i = std::prev(topologies.lower_bound(low_ts))`
2. if you meant *lower or equal* (which the code does now), you could do  `i = std::prev(topologies.upper_bound(low_ts))`